x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -856,21 +859,35 @@ protected Expression regexToEquals(RegexMatch<?> regexMatch, Literal literal) {
         }
     }
 
-    private static class ReplaceFieldAttributesWithExactSubfield extends OptimizerRules.OptimizerRule<LogicalPlan> {
+    private static class ReplaceFieldAttributesWithExactSubfield extends Rule<LogicalPlan, LogicalPlan> {
 
         @Override
-        protected LogicalPlan rule(LogicalPlan plan) {
-            if (plan instanceof Filter || plan instanceof OrderBy || plan instanceof Aggregate) {
-                return plan.transformExpressionsOnly(FieldAttribute.class, ReplaceFieldAttributesWithExactSubfield::toExact);
-            }
-            return plan;
-        }
+        public LogicalPlan apply(LogicalPlan plan) {
+            Map<EsField, FieldAttribute> esFieldToAttribute = new HashMap<>();
+
+            return plan.transformUp(LogicalPlan.class, p -> {
+                // the source is skipped but used to get the mapping of fields to attributes
+                if (p instanceof EsRelation r) {
+                    for (Attribute out : r.output()) {
+                        // map the field to their actual attributes
+                        if (out instanceof FieldAttribute fa) {
+                            esFieldToAttribute.put(fa.field(), fa);
+                        }
+                    }
+                }
+                // everything else uses the map to replace the field with the proper attribute
+                else {
+                    p = p.transformExpressionsOnly(FieldAttribute.class, f -> {
+                        if (f.dataType() == DataTypes.TEXT && f.getExactInfo().hasExact()) {
```

Comment:
Why the `TEXT` check? I mean, I know there are no other field types that use the "exact" functionality, but it's a nice functionality that elegantly abstracts this concept.

Reply:
Because there can be other field types with "exact" subfields (keyword parent field with a keyword sub-field) but only `text` has a concrete use of the sub-field.